File: 2022/day2.rb
lines = File.readlines('./inputs/day2.txt')

def part1(opponent, player)
  win = (player == 'Y' && opponent == 'A') ||
    (player == 'Z' && opponent == 'B') ||
    (player == 'X' && opponent == 'C')
  draw = (player == 'X' && opponent == 'A') ||
    (player == 'Y' && opponent == 'B') ||
    (player == 'Z' && opponent == 'C')
  shape = 1
  if player == 'Y'
    shape = 2
  end
  if player == 'Z'
    shape = 3
  end
  outcome = win ? 6 : (draw ? 3 : 0)
  return shape + outcome 
end

def part2(opponent, ends)
  outcome = ends == 'Z' ? 6 : (ends == 'Y' ? 3 : 0)
  if ends == 'X'
    shape = opponent == 'A' ? 3 : (opponent == 'B' ? 1 : 2)
  elsif ends == 'Y'
    shape = opponent == 'A' ? 1 : (opponent == 'B' ? 2 : 3)
  else
    shape = opponent == 'A' ? 2 : (opponent == 'B' ? 3 : 1)
  end
  return shape + outcome
end

puts lines.map { |line| part1(*line.split(' ')) }.sum
puts lines.map { |line| part2(*line.split(' ')) }.sum

File: 2023/problems/day2.py
import sys
import re

lines = sys.stdin.readlines()

def part1():
    gameids = []
    for line in lines:
        line = line.strip()
        gameid, game, *_ = line.split(': ')
        gamenum = int(gameid.split(' ')[1])
        rounds = game.split('; ')
        invalid = False
        for r in rounds:
            red, green, blue = 12, 13, 14
            draws = r.split(', ')
            for draw in draws:
                count, color, *_ = draw.split(' ')
                if color == 'red':
                    red -= int(count)
                elif color == 'green':
                    green -= int(count)
                elif color == 'blue':
                    blue -= int(count)
                if red < 0 or green < 0 or blue < 0:
                    invalid = True
                    break
            if invalid:
                break
        if not invalid:
            gameids.append(gamenum)
    print(sum(gameids))

def part2():
    total = 0
    for line in lines:
        line = line.strip()
        gameid, game, *_ = line.split(': ')
        gamenum = int(gameid.split(' ')[1])
        rounds = game.split('; ')
        max_red, max_green, max_blue = 0, 0, 0
        for r in rounds:
            draws = r.split(', ')
            for draw in draws:
                count, color, *_ = draw.split(' ')
                if color == 'red':
                    max_red = max(max_red, int(count))
                elif color == 'green':
                    max_green = max(max_green, int(count))
                elif color == 'blue':
                    max_blue = max(max_blue, int(count))
        total += max_red * max_green * max_blue
    print(total)

part1()
part2()
